### DISCUSSION TITLE
I can't figure out the object graph

### DISCUSSION THREAD

**Discussion by karthink** [qtile/qtile#3659] [Q&A]
I've been trying to understand Qtile's API and scripting interface, but it's been frustrating. Here are some questions:

1. Is the root of the command graph a Lazy object or a Qtile object? I've seen references to both in the documentation.

2. How do I find what methods are available for graph objects? I tried

```python
from libqtile.command.client import InteractiveCommandClient
c = InteractiveCommandClient()

dir(c)
dir(c.group)
```

But there was nothing relevant in the list of methods on `c`. Since then it's been a painful process of trial and error. Following [this image](http://docs.qtile.org/en/latest/_images/graphviz-1857f9119645ba627592ff4e300e43ef06ab0012.png) in the docs, I gathered that I could follow the arrows like `c.group.layout`, `c.screen[1].widget["textbox12"].bar` etc, but I don't actually know how to find the available methods defined on any of these objects. I only know some adhoc examples, from the doc. Some of these objects have an `info` method, the group object has a `to_screen` method, etc. Specifically, it would be great if I got auto-completion in my python repl for these methods.

3. Is there some way to introspect the return value or effect of a function at runtime without modifying it, adding a logger.warn statement and restarting Qtile? This gets tiresome very fast when combined with trial and error. For example, I have this function:

```python
def window_to_previous_screen(switch_group=False, switch_screen=False):
    def inner(qtile):
        i = qtile.screens.index(qtile.current_screen)
        if i != 0:
            group = qtile.screens[i - 1].group.name
            qtile.current_window.togroup(group, switch_group=switch_group)
            if switch_screen == True:
                qtile.cmd_to_screen(i - 1)
    return lazy.function(inner)
```

Can I debug this function (the `inner` one that's actually called, not the closure constructor) and step through it at runtime? 

4. Why does the above function fail if I replace the inner def with `def inner(qtile: Qtile):`? I'm not familiar with type annotations in Python, am I doing something wrong here?

5. How do I get autocompletion (with LSP, say) of the properties/methods of `qtile` in the above function? I'm groping in the dark when trying to write some simple functions like this one, to either switch to a group (with name `name`) or to toggle the previously selected group:

```python
@lazy.function
def group_back_and_forth(qtile, name: str):
    if qtile.group[name] == qtile.current_group.name:
        qtile.cmd_toggle_group()
    else:
        qtile.group[name].toscreen()
```

I assumed I could traverse the graph structure starting from a `qtile` object much like I could with the `InteractiveCommandClient` object, but `qtile` does not have a group attribute. I had to do

```python
i = qtile.screens.index(qtile.current_screen)
group = qtile.screens[i - 1].group.name

```

The question is not how to fix this particular error, but how to figure out what's available without having to restart Qtile and look at tracebacks in the log.

(Yes, I am aware that this particular feature is available via `lazy.group[grpname].toscreen(toggle=True)`.)

6. Why does qtile have both `cmd_*` methods and methods without this prefix?

**Comment by elParaguayo** [comment]
Loooots of questions. Here we go!

> I've been trying to understand Qtile's API and scripting interface, but it's been frustrating. Here are some questions:
> 
> 1. Is the root of the command graph a Lazy object or a Qtile object? I've seen references to both in the documentation.

The command graph is separate to the qtile object. Do not think they are comparable. The idea of the command graph is to expose various parts of qtile over an interface such that it can be controlled by external scripts.

> 2. How do I find what methods are available for graph objects? I tried
> 
> ```python
> from libqtile.command.client import InteractiveCommandClient
> c = InteractiveCommandClient()
> 
> dir(c)
> dir(c.group)
> ```
> 
> But there was nothing relevant in the list of methods on `c`. Since then it's been a painful process of trial and error. Following [this image](http://docs.qtile.org/en/latest/_images/graphviz-1857f9119645ba627592ff4e300e43ef06ab0012.png) in the docs, I gathered that I could follow the arrows like `c.group.layout`, `c.screen[1].widget["textbox12"].bar` etc, but I don't actually know how to find the available methods defined on any of these objects. I only know some adhoc examples, from the doc. Some of these objects have an `info` method, the group object has a `to_screen` method, etc. Specifically, it would be great if I got auto-completion in my python repl for these methods.
> 

Yeah. That won't work. You can't use `dir` on the command graph. If you want to see available commands for a node on the graph, use the `.commands()` call.

> 3. Is there some way to introspect the return value or effect of a function at runtime without modifying it, adding a logger.warn statement and restarting Qtile? This gets tiresome very fast when combined with trial and error. For example, I have this function:
> 
> ```python
> def window_to_previous_screen(switch_group=False, switch_screen=False):
>     def inner(qtile):
>         i = qtile.screens.index(qtile.current_screen)
>         if i != 0:
>             group = qtile.screens[i - 1].group.name
>             qtile.current_window.togroup(group, switch_group=switch_group)
>             if switch_screen == True:
>                 qtile.cmd_to_screen(i - 1)
>     return lazy.function(inner)
> ```
> 
> Can I debug this function (the `inner` one that's actually called, not the closure constructor) and step through it at runtime?
> 

Sorry, not sure I follow what you're trying to do.

> 4. Why does the above function fail if I replace the inner def with `def inner(qtile: Qtile):`? I'm not familiar with type annotations in Python, am I doing something wrong here?

2 questions:
1) Why do you need to do this?
2) Did you import the `Qtile` class definition in your config? (what does the error message say?)

> 5. How do I get autocompletion (with LSP, say) of the properties/methods of `qtile` in the above function? I'm groping in the dark when trying to write some simple functions like this one, to either switch to a group (with name `name`) or to toggle the previously selected group:
> 
> ```python
> @lazy.function
> def group_back_and_forth(qtile, name: str):
>     if qtile.group[name] == qtile.current_group.name:
>         qtile.cmd_toggle_group()
>     else:
>         qtile.group[name].toscreen()
> ```
> 
> I assumed I could traverse the graph structure starting from a `qtile` object much like I could with the `InteractiveCommandClient` object, but `qtile` does not have a group attribute. I had to do
> 
> ```python
> i = qtile.screens.index(qtile.current_screen)
> group = qtile.screens[i - 1].group.name
> ```
> 
> The question is not how to fix this particular error, but how to figure out what's available without having to restart Qtile and look at tracebacks in the log.
> 
> (Yes, I am aware that this particular feature is available via `lazy.group[grpname].toscreen(toggle=True)`.)
> 

I'm guessing that autocompletion struggles here because it looks up what `qtile` could be and sees that it defaults to `None`.

> 6. Why does qtile have both `cmd_*` methods and methods without this prefix?

This is an issue that annoys me because it causes so much confusion for users.

Currently, qtile uses the `cmd_` prefix to identify those commands that are exposed on the command graph. They are then exposed without the prefix. However, if you're calling them directly (i.e. from the relevant objects) you need to call with the prefix.

There's an open PR that will fix this by removing the `cmd_` prefix.

**Reply by karthink** [reply]
> The command graph is separate to the qtile object. Do not think they are comparable. The idea of the command graph is to expose various parts of qtile over an interface such that it can be controlled by external scripts.

Okay. In that case, which should I try to use when writing functions in my config file, `lazy` or `qtile`? 

> Yeah. That won't work. You can't use `dir` on the command graph. If you want to see available commands for a node on the graph, use the `.commands()` call.

Got it, thanks. I found that `commands()` also works in qtile shell.

> > Can I debug this function (the `inner` one that's actually called, not the closure constructor) and step through it at runtime?
> 
> Sorry, not sure I follow what you're trying to do.

1. I want to run a function I wrote (like the one above) from a python REPL when qtile is running, and
2. I want to step through this function with a debugger. This way I can test functions without restarting qtile over and over.

("Live coding" like this is something I'm used to doing in Common Lisp applications (Nyxt, StumpWM) and in Emacs.)

> 2 questions:
> 
> 1. Why do you need to do this?

So that LSP knows what kind of object it is, with the hope that it will give me method completion and docstrings.

> 2. Did you import the `Qtile` class definition in your config? (what does the error message say?)

D'oh. I didn't. I had

```python
from libqtile import qtile, extension
```

in the file, and assumed this was what I needed. I added

```python
from libqtile.core.manager import Qtile
```

and with the type annotation from above, I now get method autocompletion. 👍

> I'm guessing that autocompletion struggles here because it looks up what `qtile` could be and sees that it defaults to `None`.

Yeah, this was the issue. But in general, it would help if the [docs page for scripting commands](http://docs.qtile.org/en/latest/manual/ref/commands.html) also listed the properties/attributes of the Qtile class in addition to the `cmd_*` methods. For example, qtile has attributes `qtile.current_screen`, `qtile.current_group` and `qtile.current_layout` which are very useful when writing functions in my config. But short of reading the code, there was no way to discover these.

> Currently, qtile uses the `cmd_` prefix to identify those commands that are exposed on the command graph. They are then exposed without the prefix. However, if you're calling them directly (i.e. from the relevant objects) you need to call with the prefix.
> 
> There's an open PR that will fix this by removing the `cmd_` prefix.

That would be very helpful!

Overall, I still feel like I'm groping in the dark when writing helper functions. For example, I see that Qtile has an attribute named `groups_map` that appears to map group names to groups. How can I check its value interactively? It doesn't seem like information exposed on the object graph.

**Reply by elParaguayo** [reply]
> > The command graph is separate to the qtile object. Do not think they are comparable. The idea of the command graph is to expose various parts of qtile over an interface such that it can be controlled by external scripts.
> 
> Okay. In that case, which should I try to use when writing functions in my config file, `lazy` or `qtile`?

`lazy` does not call the commands so you can't use it inside functions. It creates an object (a `LazyCall`) which is passed to the server and evaluated at a particular time. It's therefore used for key bindings and mouse callbacks.

If you're writing a function within your config file then you should access `qtile`. Also, if your function is for a layout, group etc then, instead of `lazy.function` you can use `lazy.layout.function` and the function will receive the layout object as the first argument, rather than `qtile`.

> 
> > Yeah. That won't work. You can't use `dir` on the command graph. If you want to see available commands for a node on the graph, use the `.commands()` call.
> 
> Got it, thanks. I found that `commands()` also works in qtile shell.
> 
> > > Can I debug this function (the `inner` one that's actually called, not the closure constructor) and step through it at runtime?
> > 
> > 
> > Sorry, not sure I follow what you're trying to do.
> 
> 1. I want to run a function I wrote (like the one above) from a python REPL when qtile is running, and
> 2. I want to step through this function with a debugger. This way I can test functions without restarting qtile over and over.

You can't do this in a REPL as you can't access the live `qtile` object as you're in a separate python session. If you try to import `qtile` in a REPL, its value will be `None`. This is why you need to Command Graph to access objects from different sessions.

> Overall, I still feel like I'm groping in the dark when writing helper functions. For example, I see that Qtile has an attribute named `groups_map` that appears to map group names to groups. How can I check its value interactively? It doesn't seem like information exposed on the object graph.

You can get a list of groups with the object graph:
``` python
from libqtile.command.client import InteractiveCommandClient
c = InteractiveCommandClient()

print(c.items("group"))
```
`.items(node)` lists the available items in the given node.

**Reply by karthink** [reply]
> Also, if your function is for a layout, group etc then, instead of lazy.function you can use lazy.layout.function and the function will receive the layout object as the first argument, rather than qtile.

That's very handy!

> You can't do this in a REPL as you can't access the live qtile object as you're in a separate python session. If you try to import qtile in a REPL, its value will be None. This is why you need to Command Graph to access objects from different sessions.

Okay, so the command graph is for live interaction and Qtile is for functions already defined when Qtile starts.

**Reply by karthink** [reply]
I was trying out a hook function you posted in #2651:

```python
@hook.subscribe.client_new
def disable_floating(window):
    rules = [
        Match(wm_class="mpv")
    ]
    if any(window.match(rule) for rule in rules):
        window.togroup(qtile.current_group.name)
        window.cmd_disable_floating()
```

This does not work without `window.togroup(qtile.current_group.name)`. Why is that?

**Reply by elParaguayo** [reply]
That's probably because `client_new` is called before qtile has managed the window (i.e. allocated it to a group and run the matching rules). There's a different hook, `client_managed`, which would let you remove that line but you may get some flickering as the window will be positioned by the group first.

Probably easiest if you just try it and see what works best for you.

**Reply by karthink** [reply]
Thanks. So if that line is excluded, the above hook function runs and mpv is tiled, then the `client_managed` hook runs and the window is floated again according to the default floating rules. But then why doesn't the `client_managed` hook run anyway and float the window after the above function runs?

> Probably easiest if you just try it and see what works best for you.

I''m not looking for the best fix for the mpv issue, just trying to understand what's happening under the hood.

**Reply by elParaguayo** [reply]
I'd need to look at the code more closely to remind myself what happens there - I may well have got it wrong.

**Reply by karthink** [reply]
Is there some way to find out what data is being written to the qtile socket? I was wondering if I could write some json to the socket manually (e.g. using `socat`) instead of using `qtile cmd-obj`.

**Reply by m-col** [reply]
I've been thinking this too - it would be awesome if we can do that, as `qtile cmd-obj` seems to have a bit of a slow startup. Unless that is something we can fix.

**Reply by elParaguayo** [reply]
I *think* this is possible. IIRC messages are sent as json by default. Let me take a quick look later today.

**Reply by karthink** [reply]
Yeah `qtile cmd-obj` is noticeably slower than the keybound lazycalls. I've been trying to figure out the json that gets written to the socket in this call

```python
        status, result = self._client.send((call.parent.selectors, call.name, args, kwargs))
```

in `libqtile/command/interface.py`.

The result of

```sh
qtile cmd-obj -o layout -f left
```

is the following data being sent in the argument tuple above:

```text
[('layout', None)], commands, (), {}
[('layout', None)], left, (), {}
```

If I can figure out how this gets json-encoded, it should be possible to control qtile from anything. :)

I've been trying to write a "seamless" mode for Emacs that treats Emacs windows as qtile windows for the purposes of switching and shuffling. Kinda like how folks like to set up tmux splits to be seamless with Vim panes.

**Reply by elParaguayo** [reply]
There are multiple calls with cmd-obj, but I'm hoping we can just send a single string to the socket.

**Reply by karthink** [reply]
This works:

```sh
cd ~/.cache/qtile/
ln -s 'qtilesocket.:0' 'qtiletest'
echo '[[["layout", null]], "shuffle_left", [], {}]' | socat UNIX-CONNECT:./qtiletest -
```

The symbolic link is because `socat` doesn't like filenames with colons.

This swaps the window with the main window in my layout. It's much faster than `qtile cmd-obj`!

EDIT: If there's a general schema for the messages, it's going to be amazing. I'll be able to seamlessly integrate qtile with Emacs, control it from my phone, and so on!

**Reply by elParaguayo** [reply]
Nice work.

The format is `[[selectors], command, [args], {kwargs}]` so `lazy.spawn("alacritty")` should be `[[], "spawn", ["alacritty"], {}]`.

**Reply by elParaguayo** [reply]
Also, you can use the normal socket name with socat, you just need to escape the colon in the filename.

**Reply by karthink** [reply]
Perfect, thank you. Do the selectors correspond to a path on the command graph? Can they be nested above?

I mean if this is a valid selector(s)

`[["screen", null], ["group", null]]`

i'm not sure what command to use to test this. I did

```shell
echo '[[["screen", null],["group", null]], "setlayout", ["tile"], {}]' | socat UNIX-CONNECT:./qtiletest -
```

And it worked, but I don't know if this is equivalent to `qtile.current_screen.group.setlayout(...)`.

**Reply by elParaguayo** [reply]
Yes, selectors correspond to how you navigate on the command graph and, yes, they can be nested (although you don't often do that).

The first part of the selector is the name of the node. The second is the identifier. The identifier can be `None` (`null` in JSON) which means use the default object where available (e.g. current window/group/layout).

You can chain selectors as long as the command graph allows it (refer to the arrows in the docs). For example, to get the bar that your clock widget is on, you could use `[["widget", "clock"], ["bar", null]]` - this would be equivalent to `lazy.widget["clock"].bar...`

**Reply by elParaguayo** [reply]
I wonder if it's worth adding something in the docs about this...

**Reply by karthink** [reply]
Wow, this opens up command graph interaction to any language! I'm going to go experiment a little.

**Reply by m-col** [reply]
Excellent! We should document it, but as soon as we do it becomes an official API, so we should put some thought into the structure of the messages. I say that mainly because on first glance I don't quite understand the structure of `'[[["screen", null],["group", null]], "setlayout", ["tile"], {}]'`. If we can simplify that a bit, and probably check that types are being transmitted correctly, then it'd be perfect.

**Reply by elParaguayo** [reply]
So...
- the whole request is a single array
- the first member of the array is another array containing the object selectors
  - each object selector is an array which must have two members
  - the first member is the name of the node
  - the second is the identifier (or `null` for the default)
- next is the name of the command to run, a string
- next is the args to be passed to the command. This is an array
- finally we have the kwargs in a JSON object (`{key:value}`)

For your example above:
- The selectors are `[["screen", null],["group", null]]` which means go to the default screen node and, from there, go to the default group node (i.e. the active group on the current screen)
- The command is `setlayout`
- The args are `"tile"`

This is the same as `lazy.screen.group.setlayout("tile")` or `qtile cmd-obj -o screen group -f setlayout -a tile`.

**Reply by karthink** [reply]
I've been experimenting with qtile's IPC and I discovered that when you write to the qtile socket (through `socat`'s stdin say), it doesn't execute the commands:

```shell
$ socat - UNIX-CONNECT:~/.cache/qtile/qtilesocket.:0
[[["layout", null]], "left", [], {}]

```

Nothing happens until I signal INT (Control-C) or EOF (Control-D) and terminate `socat`. Then all the json I've typed in is read and acted on. This means I have to spawn a new process or network connection for any interaction with qtile through its socket. Is this intentional? And is there some way to get qtile to process a line of json posted to the socket (with a terminating newline)?

To be sure, I tried out the same command to interact with an mpv IPC socket, and mpv reads the json and acts on it whenever there's a newline char.

**Reply by m-col** [reply]
So most of the json structure makes sense, I think the `null` got me confused, and it does seem redundant... maybe we can do away with that.

> Nothing happens until I signal INT (Control-C) or EOF (Control-D) and terminate socat. Then all the json I've typed in is read and acted on. This means I have to spawn a new process or network connection for any interaction with qtile through its socket. Is this intentional? And is there some way to get qtile to process a line of json posted to the socket (with a terminating newline)?

I don't think any of this is intentional as this isn't user-facing API, it's largely an incidentally useful implementation detail. So we can improve it so that the server processes messages on newlines, sure.

**Reply by karthink** [reply]
Oh, I see it here (`ipc.py`):

```python
        try:
            reader, writer = await asyncio.wait_for(
                asyncio.open_unix_connection(path=self.socket_path), timeout=3
            )
        except (ConnectionRefusedError, FileNotFoundError):
            raise IPCError("Could not open {}".format(self.socket_path))

        try:
            send_data = _IPC.pack(msg, is_json=self.is_json)
            writer.write(send_data)
            writer.write_eof()

            read_data = await asyncio.wait_for(reader.read(), timeout=10)
        except asyncio.TimeoutError:
            raise IPCError("Server not responding")
        finally:
            # see the note in Server._server_callback()
            writer.close()
            await writer.wait_closed()
```

You open a new connection to the socket for each data packet. I'm not sure about this -- but doesn't it make more sense to keep the connection open and avoid the overhead?

**Reply by elParaguayo** [reply]
Thanks. I noticed the EOF point too. From what I can tell, this is because of these lines:
https://github.com/qtile/qtile/blob/c16f96d5e60e0df699e39d9b0db1fe91981c1db7/libqtile/ipc.py#L228-L229

`reader` is an instance of `asyncio.StreamReader` which also has a `readline()` method (https://docs.python.org/3/library/asyncio-stream.html#asyncio.StreamReader.readline) which looks for the newline character.

I imagine making that change would break a few things elsewhere but perhaps not irreparably!

**Reply by elParaguayo** [reply]
> You open a new connection to the socket for each data packet. I'm not sure about this -- but doesn't it make more sense to keep the connection open and avoid the overhead?

That's the client code. Your client doesn't need to be connected all the time (and can't be if it's just a script).

**Reply by karthink** [reply]
> So most of the json structure makes sense, I think the null got me confused, and it does seem redundant... maybe we can do away with that.

as @elParaguayo pointed out above, the `null` is not redundant, it's part of the object selector (`[["widget","clock"]]` etc).

> I don't think any of this is intentional as this isn't user-facing API, it's largely an incidentally useful implementation detail. So we can improve it so that the server processes messages on newlines, sure.

It will make a big impact to external applications interacting with qtile in a continuous manner to not have to open a connection for each call.

> That's the client code. Your client doesn't need to be connected all the time (and can't be if it's just a script).

That makes sense. This code is only run by `qtile cmd-obj`? I thought all communication (such as lazycalls) were through the socket.

**Reply by elParaguayo** [reply]
No. lazycalls don't go through the socket. They break the command down into a similar format (node selectors, command, args) and store these values. When you press the key the server builds the correct command and executes it but nothing's sent through the socket.

**Reply by karthink** [reply]
Thanks for the clarification.

I wrote up a basic Emacs package for seamless integration with qtile: the same keys move across Emacs panes/windows and window manager windows.

It involves a lot of back and forth:

1. Qtile's key handler checks if the current window matches `wm_class=emacs`. If it doesn't it calls the appropriate lazy call.

2. If it's Emacs, qtile calls `emacsclient` with an elisp function. (This involves writing to Emacs' socket server)

3. Emacs checks if the movement in the requested direction involves leaving Emacs. If it doesn't it calls an appropriate Emacs command.

4. If it involves leaving Emacs it calls `qtile cmd-obj` (or something equivalent, see below), which writes to Qtile's socket and performs the move.

What I learnt:

1. Writing to the emacs socket is fast, window movement inside Emacs is close to instantaneous.

2. `qtile cmd-obj` is too slow to be usable. There's a lag of about a second (sometimes more) just to move focus to a neighbouring (non-Emacs) window. 

3. Calling `socat` from Emacs (instead of `qtile cmd-obj`) is faster, but there's still perceptible lag. I'm guessing this is the overhead of starting and terminating the `socat` process.

4. Using Emacs' internal API to write to the socket (instead of spinning up `socat`) is faster still, with a hint of lag.

5. Guess: Maintaining a continuous connection to the socket (and reading with `asyncio.StreamReader.readline` on the qtile side) might eliminate the lag entirely and make it fully seamless. It also opens the door to more "chatty" interaction between Qtile and applications. It's kinda fuzzy, but I'm imagining some dynamic widgets I could write that will let me use Qtile's bar as an extension of Blender and Emacs' UIs!

**Reply by m-col** [reply]
> as @elParaguayo pointed out above, the null is not redundant, it's part of the object selector ([["widget","clock"]] etc).

It's needed currently based on the implementation, but the implementation could be modified so that `[[["layout", null]], "left", [], {}]` instead becomes `[[["layout"]], "left", [], {}]`. Or would that not be unambiguous?

Just thinking about the messages, each command is composed of a sequence of nodes, optionally connected via item-getters, and terminating in a function call with 0 or more arguments or keyword arguments. What is the most simple and intuitive way such a message can be constructed? Some ideas:

1. `[[["group", 0], ["layout", null]], "left", [], {}]` (current)
2. `[[["group", 0], ["layout"]], "left", [], {}]` - dropped nulls
3. `[[["group", 0], "layout"], "left", [], {}]` - dropped nulls and list when a node has no item getter
4. `[["group", [0], "layout"], "left", [], {}]` - if a path item is a list, if it's an item getter for the previous node
5. `[["group", 0, "layout"], "left", [], {}]` - if a path item is not one of `["bar", "group", "layout", "screen", "widget", "window" ,"core"]`, it's an item getter (caveat: doesn't allow item getters with those values, e.g. can't select `widget["group"]`)
6. `[["group", 0, "layout"], "left"]` - drop args when absent
7. `["group", 0, "layout", "left"]` - drop inner list, as the final element is always a function so this is unambiguous. `["group", "layout"]` would be equivalent to `lazy.group.layout()`.

What do people think? Another consideration might be if the user wants to send multiple messages at once, i.e. a list where each element is a message as constructed above. This would probably be a redundant feature if messages can be transmitted via a persistent socket that uses line breaks to delimit messages.

Personally I think dropping as much as possible is preferable, though probably not the item-getter list, so a combination of 4 and 7: `["group", [0], "layout", "left"]`

**Reply by elParaguayo** [reply]
Need to think about it, not sure all of those ideas work. Number 6 looks interesting although I'm concerned that there may be some edge cases where a node could be parsed incorrectly.

You could also do a JSON object:
```
{
    "node": ["group", 0, "layout"],
    "command": "left",
    # args and kwargs can be optional
}
```
It's a few more bytes to send but may remove some ambiguity.
